filebeat/crawler/crawler.go

```diff
@@ -23,27 +24,50 @@ import (
 */
 
 type Crawler struct {
-	// Registrar object to persist the state
-	Registrar   *Registrar
-	prospectors []*prospector.Prospector
-	wg          sync.WaitGroup
+	registrar         *registrar.Registrar
+	prospectors       []*prospector.Prospector
+	wg                sync.WaitGroup
+	spooler           *spooler.Spooler
+	prospectorConfigs []*common.Config
 }
 
-func (c *Crawler) Start(prospectorConfigs []*common.Config, eventChan chan *input.FileEvent) error {
+func NewCrawler(registrar *registrar.Registrar, spooler *spooler.Spooler, prospectorConfigs []*common.Config) (*Crawler, error) {
 
 	if len(prospectorConfigs) == 0 {
-		return fmt.Errorf("No prospectors defined. You must have at least one prospector defined in the config file.")
+		return nil, fmt.Errorf("No prospectors defined. You must have at least one prospector defined in the config file.")
 	}
 
-	logp.Info("Loading Prospectors: %v", len(prospectorConfigs))
+	return &Crawler{
+		registrar:         registrar,
+		spooler:           spooler,
+		prospectorConfigs: prospectorConfigs,
+	}, nil
+}
+
+func (c *Crawler) Start() error {
+
+	// Start the registrar
+	c.registrar.Start()
+
+	// Starting spooler
+	c.spooler.Start()
```

Comment:
let's move 'Start' back into `(*Filebeat).Run`. The crawler should not care about initializing the registrar. I would even consider not passing the registrar and spooler itself, but using some interface for abstracting away the presence of spooler/registrar.

Is registrar even needed by crawl beside 'setup' ?